oauth/src/lib.rs

```diff
@@ -162,8 +164,178 @@ fn get_socket_address(redirect_uri: &str) -> Option<SocketAddr> {
     None
 }
 
+/// Struct that handle obtaining and refreshing of access tokens
+/// Should not be instantiate by itself, use OAuthClientBuilder instead.
+pub struct OAuthClient {
+    scopes: Vec<String>,
+    redirect_uri: String,
+    should_open_url: bool,
+    message: String,
+    client: BasicClient,
+}
+
+impl OAuthClient {
+    /// Obtain a Spotify access token using the authorization code with PKCE OAuth flow.
+    /// The `redirect_uri` must match what is registered to the client ID.
+    pub async fn get_access_token(&self) -> Result<OAuthToken, OAuthError> {
+        let (pkce_challenge, pkce_verifier) = PkceCodeChallenge::new_random_sha256();
+
+        // Generate the full authorization URL.
+        // Some of these scopes are unavailable for custom client IDs. Which?
+        let request_scopes: Vec<oauth2::Scope> =
+            self.scopes.iter().map(|s| Scope::new(s.into())).collect();
+        let (auth_url, _) = self
+            .client
+            .authorize_url(CsrfToken::new_random)
+            .add_scopes(request_scopes)
+            .set_pkce_challenge(pkce_challenge)
+            .url();
+
+        if self.should_open_url {
+            open::that_in_background(auth_url.as_str());
+        } else {
+            println!("{}", auth_url);
+        }
+
+        let code = match get_socket_address(&self.redirect_uri) {
+            Some(addr) => get_authcode_listener(addr, self.message.clone()),
+            _ => get_authcode_stdin(),
+        }?;
+        trace!("Exchange {code:?} for access token");
+
+        let resp = self
+            .client
+            .exchange_code(code)
+            .set_pkce_verifier(pkce_verifier)
+            .request_async(async_http_client)
+            .await;
+
+        let token = resp.map_err(|e| OAuthError::ExchangeCode { e: e.to_string() })?;
+        trace!("Obtained new access token: {token:?}");
+
+        let token_scopes: Vec<String> = match token.scopes() {
+            Some(s) => s.iter().map(|s| s.to_string()).collect(),
+            _ => self.scopes.clone(),
+        };
+        let refresh_token = match token.refresh_token() {
+            Some(t) => t.secret().to_string(),
+            _ => "".to_string(),
+        };
```

Comment:
Shouldn't this be rather an `Option<String>` then an empty string? But I suppose this would break the api?

Reply:
In reality there's always a secret so I didn't see the point adding an extra layer of indirection for no benefit.